### PR TITLE
Add words per user per day chart w/progress bar

### PR DESCRIPTION
**Pull request by fuencui**
Add a Line Chart demonstrate Words collected Per User Per day at Statistic page
Add a linear progress bar shows percentage of the domains are collected and Total numbers of domains and words collected

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/TheCombine/1937)

### CODE REVIEW

**Review by jasonleenaylor — CHANGES_REQUESTED** [review]
> Reviewed 1 of 24 files at r1, 2 of 12 files at r4, 4 of 12 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/TheCombine/1937)* status: 11 of 28 files reviewed, 6 unresolved discussions (waiting on @fuencui)

___
*[`Backend/Controllers/StatisticsController.cs` line 89 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQbJYB5I5t_Y27UinW:-NQQbJYB5I5t_Y27UinX:b-o72u3m) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Controllers/StatisticsController.cs#L89)):*
> ```Smalltalk
>             }
> 
>             // var list = await _staService.GetWordsPerDayUserChartJSCounts(projectId);
> ```

You will need to handle an empty project either here, or in the front end.



_Code quote:_
```Smalltalk
            // var list = await _staService.GetWordsPerDayUserChartJSCounts(projectId);
            // if (list is null)
            // {
            //     return NoContent();
            // }
```
___
*[`Backend/Models/Statistic.cs` line 12 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQf9qj7gU4I3n0kIAF:-NQQf9qj7gU4I3n0kIAG:b-5hkjl0) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L12)):*
> ```Smalltalk
> {
> 
>     /*
> ```

These comments should be part of the doc style comments. 
I would put it after the summary like this: 
```<remarks>This object is for use in the Statistics service and is not stored in the database.</remarks>```



_Code quote:_
```Smalltalk
    /*
        data object just for Service method used only
        Not for MongoDB store
    */
```
___
*[`Backend/Models/Statistic.cs` line 16 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQfj123zwkGyksXMFp:-NQQfj123zwkGyksXMFq:b4rvkc6) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L16)):*
> ```Smalltalk
>         Not for MongoDB store
>     */
>     /// <summary> Contains Id/username, DomainHashSet/DomainCount and WordCount for Words Per User Statistics </summary>
> ```

Describing the properties makes a comment likely to become inaccurate. If someone adds a property they may forget the comment.
A better description would be:
'This class contains all the data needed by the Statistics service to report on how many words a user has collected in each domain.'



_Code quote:_
```Smalltalk
Contains Id/username, DomainHashSet/DomainCount and WordCount for Words Per User Statistics
```
___
*[`Backend/Models/Statistic.cs` line 73 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQgV_PDYsaF_MmChQA:-NQQgV_PDYsaF_MmChQB:b-7u6eju) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L73)):*
> ```Smalltalk
>         Not for MongoDB store
>     */
>     //The data structure is for per user per day chartJS use only
> ```

I think we should avoid mentioning chartJS specifically, in the future we could use a different chart library. Instead mention the service that the model supports.



_Code quote:_
```Smalltalk
chartJS
```
___
*[`Backend/Models/Statistic.cs` line 120 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQhYTW3PrGvHL-bxfO:-NQQhYTW3PrGvHL-bxfP:bx450kc) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L120)):*
> ```Smalltalk
>     {
>         [Required]
>         public string Label { get; set; }
> ```

There are several things here that I think can be improved. Labels appears to be duplicated between the list and the datasets. The Dataset contains many properties that the backend does not use. I think in your refactor too much has moved to the backend.

I think the API should probably return a list of Datasets, where the data set contained only the label and the list of integers. The frontend can compose this type into a new type and add the presentation data (Border, and Background color).

On the frontend it is a simple map operation to retrieve the list of labels from the Dataset. 



_Code quote:_
```Smalltalk
    public class ChartJsRootData
    {
        [Required]
        public List<string> Labels { get; set; }
        [Required]
        public List<Dataset> Datasets { get; set; }

        public ChartJsRootData()
        {
            Labels = new List<string>();
            Datasets = new List<Dataset>();
        }
    }

    //interface DatasetsProps {
    //   label: string;
    //   data: Array<number>;
    //   borderColor: string;
    //   backgroundColor: string;
    // }

    public class Dataset
    {
        [Required]
        public string Label { get; set; }
        [Required]
        public List<int> Data { get; set; }

        [Required]
        public string BorderColor { get; set; }

        [Required]
        public string BackgroundColor { get; set; }

        public Dataset(string label, int data)
        {
            Label = label;
            Data = new List<int>() { data };
            BorderColor = "";
            BackgroundColor = "";
        }
    }
```
___
*[`Backend/Services/StatisticsService.cs` line 61 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQjBh32fC1PA1fTOik:-NQQjBh32fC1PA1fTOil:b-lcwmtv) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Services/StatisticsService.cs#L61)):*
> ```Smalltalk
>         /// Get the count of words per day per user as a list of WordsPerDayUserChartJSCount objects <see cref="WordsPerDayUserChartJSCount"/>
>         /// </summary>
>         public async Task<List<WordsPerDayUserChartJSCount>> GetWordsPerDayUserChartJSCounts(string projectId)
> ```

ChartJS should be dropped from the type and method names.



_Code quote:_
```Smalltalk
        /// <summary>
        /// Get the count of words per day per user as a list of WordsPerDayUserChartJSCount objects <see cref="WordsPerDayUserChartJSCount"/>
        /// </summary>
        public async Task<List<WordsPerDayUserChartJSCount>> GetWordsPerDayUserChartJSCounts(string projectId)
```

**Comment by fuencui** [comment]
> > Reviewed 1 of 24 files at r1, 2 of 12 files at r4, 4 of 12 files at r5, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/sillsdev/TheCombine/1937)_ status: 11 of 28 files reviewed, 6 unresolved discussions (waiting on @fuencui)
> 
> _[`Backend/Controllers/StatisticsController.cs` line 89 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQbJYB5I5t_Y27UinW:-NQQbJYB5I5t_Y27UinX:b-o72u3m) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Controllers/StatisticsController.cs#L89)):_
> 
> > ```smalltalk
> >             }
> > 
> >             // var list = await _staService.GetWordsPerDayUserChartJSCounts(projectId);
> > ```
> 
> You will need to handle an empty project either here, or in the front end.
> 
> _Code quote:_
> 
> ```smalltalk
>             // var list = await _staService.GetWordsPerDayUserChartJSCounts(projectId);
>             // if (list is null)
>             // {
>             //     return NoContent();
>             // }
> ```
> 
> _[`Backend/Models/Statistic.cs` line 12 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQf9qj7gU4I3n0kIAF:-NQQf9qj7gU4I3n0kIAG:b-5hkjl0) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L12)):_
> 
> > ```smalltalk
> > {
> > 
> >     /*
> > ```
> 
> These comments should be part of the doc style comments. I would put it after the summary like this: `<remarks>This object is for use in the Statistics service and is not stored in the database.</remarks>`
> 
> _Code quote:_
> 
> ```smalltalk
>     /*
>         data object just for Service method used only
>         Not for MongoDB store
>     */
> ```
> 
> _[`Backend/Models/Statistic.cs` line 16 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQfj123zwkGyksXMFp:-NQQfj123zwkGyksXMFq:b4rvkc6) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L16)):_
> 
> > ```smalltalk
> >         Not for MongoDB store
> >     */
> >     /// <summary> Contains Id/username, DomainHashSet/DomainCount and WordCount for Words Per User Statistics </summary>
> > ```
> 
> Describing the properties makes a comment likely to become inaccurate. If someone adds a property they may forget the comment. A better description would be: 'This class contains all the data needed by the Statistics service to report on how many words a user has collected in each domain.'
> 
> _Code quote:_
> 
> ```smalltalk
> Contains Id/username, DomainHashSet/DomainCount and WordCount for Words Per User Statistics
> ```
> 
> _[`Backend/Models/Statistic.cs` line 73 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQgV_PDYsaF_MmChQA:-NQQgV_PDYsaF_MmChQB:b-7u6eju) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L73)):_
> 
> > ```smalltalk
> >         Not for MongoDB store
> >     */
> >     //The data structure is for per user per day chartJS use only
> > ```
> 
> I think we should avoid mentioning chartJS specifically, in the future we could use a different chart library. Instead mention the service that the model supports.
> 
> _Code quote:_
> 
> ```smalltalk
> chartJS
> ```
> 
> _[`Backend/Models/Statistic.cs` line 120 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQhYTW3PrGvHL-bxfO:-NQQhYTW3PrGvHL-bxfP:bx450kc) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L120)):_
> 
> > ```smalltalk
> >     {
> >         [Required]
> >         public string Label { get; set; }
> > ```
> 
> There are several things here that I think can be improved. Labels appears to be duplicated between the list and the datasets. The Dataset contains many properties that the backend does not use. I think in your refactor too much has moved to the backend.
> 
> I think the API should probably return a list of Datasets, where the data set contained only the label and the list of integers. The frontend can compose this type into a new type and add the presentation data (Border, and Background color).
> 
> On the frontend it is a simple map operation to retrieve the list of labels from the Dataset.
> 
> _Code quote:_
> 
> ```smalltalk
>     public class ChartJsRootData
>     {
>         [Required]
>         public List<string> Labels { get; set; }
>         [Required]
>         public List<Dataset> Datasets { get; set; }
> 
>         public ChartJsRootData()
>         {
>             Labels = new List<string>();
>             Datasets = new List<Dataset>();
>         }
>     }
> 
>     //interface DatasetsProps {
>     //   label: string;
>     //   data: Array<number>;
>     //   borderColor: string;
>     //   backgroundColor: string;
>     // }
> 
>     public class Dataset
>     {
>         [Required]
>         public string Label { get; set; }
>         [Required]
>         public List<int> Data { get; set; }
> 
>         [Required]
>         public string BorderColor { get; set; }
> 
>         [Required]
>         public string BackgroundColor { get; set; }
> 
>         public Dataset(string label, int data)
>         {
>             Label = label;
>             Data = new List<int>() { data };
>             BorderColor = "";
>             BackgroundColor = "";
>         }
>     }
> ```
> 
> _[`Backend/Services/StatisticsService.cs` line 61 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQjBh32fC1PA1fTOik:-NQQjBh32fC1PA1fTOil:b-lcwmtv) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Services/StatisticsService.cs#L61)):_
> 
> > ```smalltalk
> >         /// Get the count of words per day per user as a list of WordsPerDayUserChartJSCount objects <see cref="WordsPerDayUserChartJSCount"/>
> >         /// </summary>
> >         public async Task<List<WordsPerDayUserChartJSCount>> GetWordsPerDayUserChartJSCounts(string projectId)
> > ```
> 
> ChartJS should be dropped from the type and method names.
> 
> _Code quote:_
> 
> ```smalltalk
>         /// <summary>
>         /// Get the count of words per day per user as a list of WordsPerDayUserChartJSCount objects <see cref="WordsPerDayUserChartJSCount"/>
>         /// </summary>
>         public async Task<List<WordsPerDayUserChartJSCount>> GetWordsPerDayUserChartJSCounts(string projectId)
> ```

handle an empty project case before GetLineChartRootData called

File: Backend/Controllers/StatisticsController.cs [hunk 1]
```diff
@@ -46,6 +46,30 @@ public async Task<IActionResult> GetSemanticDomainCounts(string projectId, strin
...
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(List<WordsPerDayUserChartJSCount>))]
+        public async Task<IActionResult> GetWordsPerDayUserChartJSCounts(string projectId)
+        {
+            if (!await _permissionService.HasProjectPermission(HttpContext, Permission.WordEntry))
+            {
+                return Forbid();
+            }
+
+            // Ensure project exists.
+            var proj = await _projRepo.GetProject(projectId);
```
Comment (fuencui) [inline Backend/Controllers/StatisticsController.cs:64]: handle an empty project case before `_staService.GetLineChartRootData(projectId)`
GetLineChartRootData also handle null or empty list

**Review by jasonleenaylor — CHANGES_REQUESTED** [review]
This is looking very good. I think I have just one more suggestion.

> Reviewed 5 of 24 files at r1, 1 of 4 files at r3, 1 of 12 files at r4, 13 of 13 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-:-NQVn-ij18IKKZI7Yd2h:bit3phw)* status: 27 of 28 files reviewed, 1 unresolved discussion (waiting on @fuencui)

___
*[`Backend/Models/Statistic.cs` line 120 at r5](https://reviewable.io/reviews/sillsdev/TheCombine/1937#-NQQhYTW3PrGvHL-bxfO:-NQVl64JETAoY-w_LxXQ:b-fdi99) ([raw file](https://github.com/sillsdev/TheCombine/blob/e4429d644cb570c9d312f19cd3af0e847412a814/Backend/Models/Statistic.cs#L120)):*
<details><summary><i>Previously, fuencui (fuencui) wrote…</i></summary><blockquote>

BorderColor and BackgroundColor are been removed
Labels in the ChartRootData and Label in the Dataset have similar names but it represent different content
The Labels represent a List string of date // 01/01/2023, 02/01/2023
The Label represent UserName
</blockquote></details>

Ok, I understand now. 

The naming here is very generic which is good if we are using it for more than one chart. Because this class is used for only one chart I would prefer that we make the names here more clear, even if we have to write a little code on the front-end to map the data into the chartjs model.

So UserNames, and Date instead of Labels and Label

In the future if we find another place to use this data type we can choose generic names at that time.

**Review by fuencui — COMMENTED** [review]
> Reviewed 8 of 24 files at r1, 1 of 1 files at r2, 1 of 4 files at r3, 3 of 12 files at r4, 8 of 13 files at r6, 5 of 5 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/thecombine/1937)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @fuencui)

**Pull request merged**